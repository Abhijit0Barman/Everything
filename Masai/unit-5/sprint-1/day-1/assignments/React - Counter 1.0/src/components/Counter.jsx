import React, { useState } from "react"

const Counter = ({ initCount }) => {
    const [count, setCount] = useState(initCount)

    const handleAdd = () => {
        setCount(count + 1)
    }
    const handleReduce = () => {
        setCount(count - 1)
    }
    const handleDouble = () => {
        setCount(count * 2);
    };
    return (
        <div>
            <h2 data-testid="counter-header">Counter</h2>
            <h3 data-testid="count">{count}</h3>
            <button data-testid="add-btn" onClick={handleAdd}>+</button>
            <button data-testid="subtract-btn" onClick={handleReduce}>-</button>
            <button data-testid="double-btn" onClick={handleDouble}>Double</button>
        </div>
    )
}

export default Counter