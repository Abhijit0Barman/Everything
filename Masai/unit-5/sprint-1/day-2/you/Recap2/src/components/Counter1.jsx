import React, { useState } from "react";

export const Counter1 = () => {
  const state = useState(0);
  const count = state[0];
  const setCount = state[1];

  return (
    <div>
      <h2>Counter1: {count}</h2>
      <button onClick={() => setCount((p) => p + 1)}>Add</button>
    </div>
  );
};
