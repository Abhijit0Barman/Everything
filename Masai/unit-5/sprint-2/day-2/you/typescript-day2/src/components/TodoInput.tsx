import React, { useState } from "react";
import { Todo } from "../constants";
import { addTodo } from "../api";

interface TodoInputProps {/*   
    // handleUpdate: () => void  //for multiple user    //this is not taking or passing anything, thats why "void"
    its similar to type but no need "(=) equal-to"
    interface we can only use for objects, 
    Not for primitive data-types
    */
    handleUpdate: (res: Todo) => void;  /*  but this is taking value as a parameter,
    so also have to give "type" of that value-parameter and return nothing so "void"  */
}


export const TodoInput = ({ handleUpdate }: TodoInputProps) => {
    const [title, setTitle] = useState<string>("");

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTitle(e.target.value);
    };

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        //adding "type" from (constants.ts) after importing
        const newTodo: Todo = {
            title,
            status: false,
        };

        addTodo(newTodo) //calling post_request from (api.ts) after importing
            .then((res) => {
                // handleUpdate();  
                /* this is for (Multiple-User)
                handleUpdate()  will get (call/trigger) only after successful post-request, 
                then return response will trigger the function  */
                handleUpdate(res); /*
                in this case, after getting response back which is post-request Data,
                then I am passing the data as a parameter with handleUpdate(res) function
                */
            });
        setTitle("");
    };

    return (
        <div>
            <form onSubmit={handleSubmit}>
                <input type="text" onChange={handleChange} value={title} />
                <button type="submit">Add Todo</button>
            </form>
        </div>
    );
};
