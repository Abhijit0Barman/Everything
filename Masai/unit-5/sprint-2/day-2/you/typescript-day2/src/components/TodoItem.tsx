import React from "react";
import { Todo } from "../constants";
import { toggleTodo } from "../api";
import { deleteTodoById } from "../api";

interface TodoProps extends Todo {
  //we can inherit property from different "types" using "extends"
  handleStatusUpdate: (parameter: Todo) => void;
  handleDeleteUpdate: (x: number) => void;
}

// const TodoItem = ({ id, title, status }: Todo) => {//here-also adding type from (constants.ts)
const TodoItem = ({
  id,
  title,
  status,
  handleStatusUpdate,
  handleDeleteUpdate,
}: TodoProps) => {
  /* taking 1more parameter which is not present in "type" Todo from (constants.ts), 
  thats why creating a new "interface" by inheriting old "type" */

  const handleTogggle = () => {
    /* if (id) {
        toggleTodo(id, !status);
      }
    */

    toggleTodo(!status, id) //PATCH-REQUEST
      .then((res) => {
        // after getting response, it will trigger
        handleStatusUpdate(res);
      });
  };

  const handleDelete = () => {
    /* if (id) {
        deleteTodoById(id);
      }
    */

    deleteTodoById(id) //DELETE-REQUEST
      .then((res) => {
        // after getting response, it will trigger
        // console.log(res);//200

        if (res === 200) {
          if (id) {
            handleDeleteUpdate(id);
            // console.log(`delete s`);
          }
        }
      });
  };

  return (
    <div>
      <h3>
        {title} -- status: {status ? "True" : "False"}
      </h3>
      <button onClick={handleTogggle}>TOGGLE</button>
      <button onClick={handleDelete}>DELETE</button>
    </div>
  );
};

export default TodoItem;
