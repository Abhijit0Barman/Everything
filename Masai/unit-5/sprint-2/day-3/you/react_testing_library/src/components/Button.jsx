import React from "react";

export const Button = ({ children, func }) => {
  return (
    <button data-testid="test-button" onClick={func}>
      {children}
    </button>
  );
};
