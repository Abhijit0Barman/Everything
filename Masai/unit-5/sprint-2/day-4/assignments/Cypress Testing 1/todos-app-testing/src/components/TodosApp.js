import React, { useState, useRef } from 'react';

function TodosApp() {
  const [todo, setTodo] = useState('');
  const [todos, setTodos] = useState([]);
  const todoInputRef = useRef(null);

  const handleTodoChange = (event) => {
    setTodo(event.target.value);
  };

  const handleAddTodo = () => {
    setTodos([...todos, todo]);
    setTodo('');
  };

  return (
    <div>
      <h1>Todos</h1>
      <input
        className="todo-input"
        type="text"
        placeholder="Add a new todo"
        value={todo}
        onChange={handleTodoChange}
        ref={todoInputRef}
      />
      <button onClick={handleAddTodo}>Add Todo</button>
      <ul>
        {todos.map((item, index) => (
          <li key={index}>{item}</li>
        ))}
      </ul>
    </div>
  );
}

export default TodosApp;
