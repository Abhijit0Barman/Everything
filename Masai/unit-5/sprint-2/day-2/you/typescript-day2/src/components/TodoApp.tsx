import React, { useEffect, useState } from "react";
import { TodoInput } from "./TodoInput";
import { Todo } from "../constants";
import { getTodo } from "../api";
import TodoItem from "./TodoItem";

const TodoApp = () => {
  const [todos, setTodos] = useState<Todo[]>([]); //adding type"<Todo[]>" from->constants.ts
  // const [trigger, setTrigger] = useState<boolean>(false); //adding a type "<boolean>"    //for multiple user

  useEffect(() => {
    getTodo() //calling get-request from (api.ts) and returning a promise
      .then((res) => {
        // getting res.data which is returned from the getTodo() -> (api.ts)
        setTodos(res);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);
  /* }, [trigger]);   //for multiple user
  //it will cz re-render after post request by using handleUpdate function

  const handleUpdate = () => { //this function pass to the component, where is "input taking"   //for multiple user
    setTrigger(prev => !prev)  //only toggle the value for causing Re-Render
  }
  */

  const handleUpdate = (newTodo: Todo) => {
    /*
    after passing the function, wherever the calling this function, it will take object as a parameter which is new post-request Data,    
    thats why we also have to give, type of the data by importing from (constants.ts)
    in this case we are not calling/triggr the get-request,
    instead,
    we are updating the state with the new data,
    like (...spreading old-data,new-data) this
    setTodos((prev) => {
      return [...prev, newTodo]
    });
    OR
    down below in 1 line. */
    setTodos((prev) => [...prev, newTodo]); //adding data manually rather then getRequest
  };

  const handleStatusUpdate = (res: Todo) => {    //⭐⭐⭐⭐
    setTodos((prev) => {
      return prev.map((el) => (el.id === res.id ? res : el));
      //COMPARING (PREV-ELEMENT.ID VS UPDATED-RESPONSE.ID)
    });
  };
  /*
const handleStatusUpdate = (id: number) => {              //⭐⭐⭐⭐
    setTodos((prev) => {
      return prev.map((el) => (el.id === id ? {...el, status: !el.status} : el));
    })
  }
  */

  // delete-request in response only give you Empty-Object.
  // So thats why you have to use id & filter
  const handleDeleteUpdate = (x: number) => {    //⭐⭐⭐⭐
    setTodos((prev) => {
      return prev.filter((el) => el.id !== x);
    });
  };

  return (
    <div>
      {/* 
      it will show error after passing the handle function, 
      Un-till you de-structure in input taking component, 
      you also have to give the interface, 
      by creating interface with return type in this case "void",
      cz this 1ST TYPE "Not taking parameter nor returning anything".
      */}
      <TodoInput handleUpdate={handleUpdate} />
      {todos?.map((item) => (
        // <TodoItem {...item} key={item.id} />
        <TodoItem
          {...item}
          key={item.id}
          handleStatusUpdate={handleStatusUpdate}
          handleDeleteUpdate={handleDeleteUpdate}
        />
      ))}
    </div>
  );
};

export default TodoApp;
