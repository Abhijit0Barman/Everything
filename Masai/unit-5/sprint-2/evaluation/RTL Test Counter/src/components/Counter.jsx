import React, { useState } from "react";
import { Button } from "./Button";

export const Counter = () => {
  const [count, setCount] = useState(0);

  return (
    <div>
      <h1> {count}</h1>
      <Button>Add</Button>
      <Button>Reduce</Button>
    </div>
  );
};
