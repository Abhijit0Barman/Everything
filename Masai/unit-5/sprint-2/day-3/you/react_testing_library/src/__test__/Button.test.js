// RTL has  its own virtual rendering environment
// ðŸ”¶Button Component
/*
 * whether Button is rendered on Dom or not.
 * Text content should match.
 * Should call the function when clicked?
 */

import { fireEvent, render, screen } from "@testing-library/react";
import { Button } from "../components/Button";
// import { click } from "@testing-library/user-event/dist/click";
import renderer from "react-test-renderer";

describe("Testing Button Component", () => {
  it("Button should be renderd on DOM", () => {
    render(<Button>Click Me</Button>);
    const button = screen.getByText("Click Me");
    // screen.debug();
    expect(button).toBeInTheDocument();
  });

  it("Check if the button is have correct text or not", () => {
    render(<Button>Test</Button>);
    const button = screen.getByTestId("test-button");
    expect(button).toHaveTextContent("Test");
    render(<Button>Another</Button>);
    const button2 = screen.getAllByTestId("test-button");
    expect(button2[1]).toHaveTextContent("Another"); //regx
  });

  it("Should execute the function when clicked", () => {
    const mockFunc = jest.fn();
    render(<Button func={mockFunc}>Click Me</Button>);
    const button = screen.getByTestId("test-button");
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    // expect(mockFunc).toBeCalled();
    expect(mockFunc).toBeCalledTimes(4);
  });

  it("Should match the snapshot", () => {
    const tree = renderer.create(<Button>Click Me</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
