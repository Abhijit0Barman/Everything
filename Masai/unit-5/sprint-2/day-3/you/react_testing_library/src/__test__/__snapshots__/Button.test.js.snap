// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button Component Should match the snapshot 1`] = `
<button
  data-testid="test-button"
>
  Click Me
</button>
`;
